lib/src/types.rs

```diff
@@ -545,3 +570,436 @@ pub struct TerraformStateRoles {
+                                        node.name = inst.name.clone()
+                                    };
+                                }
+                            }
+
+                            node
+                        })
+                        .collect();
+
+                    result.append(&mut nodes);
+                }
+
+                return Ok((result, state.s3_cache));
+            }
+        }
+    }
+}
+
+type NomadClients = Vec<NomadClient>;
+type NomadAllocs = Vec<NomadAlloc>;
+type BitteNodes = Vec<BitteNode>;
+pub type ClusterHandle = JoinHandle<anyhow::Result<BitteCluster>>;
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+#[serde(untagged)]
+pub enum AllocIndex {
+    Int(u32),
+    String(String),
+}
+
+impl AllocIndex {
+    pub fn get(&self) -> Option<u32> {
+        match self {
+            Self::Int(i) => Some(*i),
+            Self::String(_) => None,
+        }
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct NomadAlloc {
+    #[serde(rename = "ID")]
+    pub id: Uuid,
+    #[serde(rename = "JobID")]
+    pub job_id: String,
+    #[serde(rename = "Namespace")]
+    pub namespace: String,
+    #[serde(rename = "TaskGroup")]
+    pub task_group: String,
+    #[serde(rename = "ClientStatus")]
+    pub status: String,
+    #[serde(
+        rename(deserialize = "Name", serialize = "Index"),
+        deserialize_with = "pull_index"
+    )]
+    #[serde(alias = "Index")]
+    pub index: AllocIndex,
+    #[serde(rename = "NodeID")]
+    pub node_id: Uuid,
+}
+
+impl NomadAlloc {
+    async fn find_allocs(client: Arc<Client>, domain: String) -> anyhow::Result<NomadAllocs> {
+        let allocs = client
+            .get(format!("https://nomad.{}/v1/allocations", domain))
+            .query(&[("namespace", "*"), ("task_states", "false")])
+            .send()
+            .await?
+            .json::<NomadAllocs>()
+            .await?;
+        Ok(allocs)
+    }
+}
+
+fn pull_index<'de, D>(deserializer: D) -> Result<AllocIndex, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    let buf = AllocIndex::deserialize(deserializer)?;
+
+    match buf {
+        AllocIndex::Int(i) => return Ok(AllocIndex::Int(i)),
+        AllocIndex::String(s) => {
+            let search = Regex::new("[0-9]*\\]$").unwrap().find(&s).unwrap().as_str();
+
+            let index = &search[0..search.len() - 1];
+            let index: u32 = index.parse().unwrap();
+
+            return Ok(AllocIndex::Int(index));
+        }
+    }
+}
+
+impl BitteCluster {
+    pub async fn new() -> anyhow::Result<Self> {
+        let name = env::var("BITTE_CLUSTER")?;
+        let domain = env::var("BITTE_DOMAIN")?;
+        let provider: BitteProvider = {
+            let string = env::var("BITTE_PROVIDER")?;
+            match string.parse() {
+                Ok(v) => Ok(v),
+                Err(_) => Err(Error::ProviderError { provider: string }),
+            }?
```

Comment:
Touching of ENV, and validation should probably occur outside of a constructor call, generally you just want to create internally consistent logic, not have a bunch of business logic :)

Then again, if I were to re-write this, It would probably be named just slightly different and do the same thing, so I'm not sure.

Reply:
Fair point. My first implementation actually did pass the environment in, but it added a function simply to call the constructor, which I felt was kinda overkill, and I decided to add it in, because the struct cannot be built without this information and so seems like it's not too horrible here.